### ISSUE TITLE
[Feature Request] Trying to get Amp to use the hosts docker

### ISSUE THREAD

**Issue by Joly0** [MitchTalmadge/AMP-dockerized#100]
Its possible for containers (such as portainer) to hook into the hosts docker engine through a docker.sock. Using that, those containers are able to create more containers on the host os.

Amp already has the option to create instances in docker containers. It would be great (especially for opening ports) if it would be possible to have a master instance with this image running ads and telling it to create instances in containers using a docker.sock which creates them on the host os.

Though i am not sure, if this does work currently with amp (or ever), if some capabilities are missing or not (havent tested, this is just a suggestion for something it might be useful to look into).
In theory this could work, which would mean, we just need this image for ads (smaller size, better maintainability and more) and could use the official ones for the game instances.

**Comment by Joly0** [comment]
https://support.cubecoders.com/thread/help-on-setting-amp-in-a-docker-container/ some points where outlined here already 2 years ago (i posted my own amp container there back then lol)

**Comment by MitchTalmadge** [comment]
I have wondered the same thing. I will have to look into this

**Comment by Joly0** [comment]
I had some time tinkering with this idea though i dont get it to work as intended. 

I installed docker inside the container and added the hosts docker socket through a volume (that should in theory work) but amp was unable to create instances like that. 

I am not sure, if amp maybe completly ignores the socket and just directly uses docker itself (which might be plausible) and would result in this feature to fail, unless we come up with a proper workaround.

Other than that there are several projects like DinD (Docker-in-Docker) which sound like they are what we need, but it also sounds like a big change to the docker as it is currently.

Maybe i can further test this in the next days/weeks but this is, unfortunately, what i found

**Comment by xionous** [comment]
> I had some time tinkering with this idea though i dont get it to work as intended.
> 
> I installed docker inside the container and added the hosts docker socket through a volume (that should in theory work) but amp was unable to create instances like that.
> 
> I am not sure, if amp maybe completly ignores the socket and just directly uses docker itself (which might be plausible) and would result in this feature to fail, unless we come up with a proper workaround.
> 
> Other than that there are several projects like DinD (Docker-in-Docker) which sound like they are what we need, but it also sounds like a big change to the docker as it is currently.
> 
> Maybe i can further test this in the next days/weeks but this is, unfortunately, what i found

There is a reason it didn't work for you, it needs more then just the docker socket. a lot more.

If you want to get this to work you need to do a few things:

1. bind mount the docker socket from the host
2. bind mount a directory from the host to place instances into that is identical file path on the host and in the main AMP container. This is needed because the instance container on the host will create it's appdata in a folder on the host machine and the AMP container needs to be able to read/write info to that directory
3. add a datastore in the AMP UI that uses the aforementioned volume to store instances \*
4. make sure to select the correct datastore when you create an instance \*\*\*\*
5. change the auth URL config option in new instance defaults to be the `host.ip:port` so the outside container can reach the AMP auth end point \*\*\*
6. enable create in docker in new instance defaults \*\*\*
7. change the network config in AMP to not use the host network so that it uses the default bridge network for the outside containers \*\*
8. change the default AMP binding ip to 0.0.0.0 in network config \*\*
9. change the service port range in networking to one that is available on the host system \*\*
10. bind mount the scripts folder from the install instructions and create a files called `startup.sh` in it
11. add the following stuff to that file (it will install docker and set some things up on startup to allow the management of instances to work):
```
#!/bin/bash
# make sure change `groupmod -g 281 docker` to match the group ID of docker on your host so the group ID is identical between the container and host.
apt update; apt -y install apt-transport-https ca-certificates curl software-properties-common; curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add -; add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu focal stable"; apt -y install docker-ce; usermod -aG docker amp; groupmod -g 281 docker; apt -y install nginx
cat > /etc/nginx/sites-enabled/default <<EOF
server {
    listen 8081;

    location / {
        proxy_pass http://host-ip:8081;
    }
}
# you can create multiple copies of this block, 1 for each instance and the port it uses. the ports must match the port each instance uses. you can create as many as you think you will need, server blocks that are not used will not cause an issue.
# be sure to change the host-ip part to your hosts IP address
server {
    listen 8082;

    location / {
        proxy_pass http://host-ip:8082;
    }
}
EOF
service nginx restart
```

Some images:
\* ![image](https://user-images.githubusercontent.com/813461/215920181-42c03215-57a9-4903-af58-a9e2ccdb8838.png)
\*\* ![image](https://user-images.githubusercontent.com/813461/215920044-ea5c9e06-fb3f-44cd-936f-6ec4430f1219.png)
\*\*\* ![image](https://user-images.githubusercontent.com/813461/215920087-a2246249-d08d-41aa-8944-e0bdf19c846d.png)
\*\*\*\* ![image](https://user-images.githubusercontent.com/813461/215920203-35e7a7a5-bcdf-43ce-8c94-9092086d59c5.png)

**Comment by Dalarialus** [comment]
It may be worth looking into nestybox/sysbox for this - it's a lot more secure than DinD, and less kludgey than other workarounds. It would require changes to the base image to add things like systemd, but once implemented it should be stable and secure - it doesn't even need to be run in a privileged container!

**Comment by xionous** [comment]
> It may be worth looking into nestybox/sysbox for this - it's a lot more secure than DinD, and less kludgey than other workarounds. It would require changes to the base image to add things like systemd, but once implemented it should be stable and secure - it doesn't even need to be run in a privileged container!

My solution is not DinD it only installs Docker to use the binary to talk to the Docker socket on the host system which is bind mounted to the container. Also my solution does not require a privileged container.

**Comment by MitchTalmadge** [comment]
I definitely love this idea and want it to work too! I am focusing on #242 but then I will shift attention to this :)

**Comment by GasPoweredTomato** [comment]
> I definitely love this idea and want it to work too! I am focusing on [#242](https://github.com/MitchTalmadge/AMP-dockerized/issues/242) but then I will shift attention to this :)

Is this issue being worked? I'm coming from Pterodactyl which had this functionality and would love to see it on AMP too. I'm not a fan of running a debian VM on unraid just to use AMP.

**Comment by Joly0** [comment]
> > I definitely love this idea and want it to work too! I am focusing on [#242](https://github.com/MitchTalmadge/AMP-dockerized/issues/242) but then I will shift attention to this :)
> 
> Is this issue being worked? I'm coming from Pterodactyl which had this functionality and would love to see it on AMP too. I'm not a fan of running a debian VM on unraid just to use AMP. 

Not sure if this is worked on currently. There has been some work in the past, but I think so far it has never been officially finished. Though wanted to tell you, that LXC is also an option on Unraid, not necessarily a VM

**Comment by GasPoweredTomato** [comment]
> > > I definitely love this idea and want it to work too! I am focusing on [#242](https://github.com/MitchTalmadge/AMP-dockerized/issues/242) but then I will shift attention to this :)
> > 
> > 
> > Is this issue being worked? I'm coming from Pterodactyl which had this functionality and would love to see it on AMP too. I'm not a fan of running a debian VM on unraid just to use AMP.
> 
> Not sure if this is worked on currently. There has been some work in the past, but I think so far it has never been officially finished. Though wanted to tell you, that LXC is also an option on Unraid, not necessarily a VM

I saw LXC was an option. I installed the plugin but haven't started using it yet. Do you use AMP in an LXC?

**Comment by Joly0** [comment]
> > > > I definitely love this idea and want it to work too! I am focusing on [#242](https://github.com/MitchTalmadge/AMP-dockerized/issues/242) but then I will shift attention to this :)
> > > 
> > > 
> > > Is this issue being worked? I'm coming from Pterodactyl which had this functionality and would love to see it on AMP too. I'm not a fan of running a debian VM on unraid just to use AMP.
> > 
> > 
> > Not sure if this is worked on currently. There has been some work in the past, but I think so far it has never been officially finished. Though wanted to tell you, that LXC is also an option on Unraid, not necessarily a VM
> 
> I saw LXC was an option. I installed the plugin but haven't started using it yet. Do you use AMP in an LXC?

Yes, been using it since LXC is an option. I used Ich777´s template for amp lxc a while, then created my own based on ubuntu, but both are outdated. Will probably update ich777´s which is based on debian in the future for myself so i can switch back to it, but have not found the time to do that

**Comment by IceOfWraith** [comment]
It honestly sounds like a lot of hoops to jump through compared to making a regular LXC or VM and running the one line installer that gives you an officially supported setup. VM is most recommended by us at CubeCoders for security and isolation reasons.

**Comment by GasPoweredTomato** [comment]
> It honestly sounds like a lot of hoops to jump through compared to making a regular LXC or VM and running the one line installer that gives you an officially supported setup. VM is most recommended by us at CubeCoders for security and isolation reasons.

Yeah it is, I just really prefer docker since it's easy to access all the files and there is less overhead than a VM. Currently running in a Debian VM.

**Comment by HaoSs07** [comment]
> It honestly sounds like a lot of hoops to jump through compared to making a regular LXC or VM and running the one line installer that gives you an officially supported setup. VM is most recommended by us at CubeCoders for security and isolation reasons.

Why would it be ? Literally all docker managers run themselves in docker.
Anyway, still wanting this feature.

**Comment by IceOfWraith** [comment]
Docker having access to the host in that way is inherently insecure, if that's what you were asking about. Just because they do it, doesn't mean it's good.

**Comment by HaoSs07** [comment]
I was talking about the " lot of hoops " thing. I'm running a few games on a server, not NASA, I'm fine with the insecurity accessing docker host trough a docker brings, i'm already doing that. My server is big enough to run a few game server but not big enough to start VM ing.  A security disclaimer when turning this on should be enough.

**Comment by IceOfWraith** [comment]
If it has enough power to run game servers then it likely can handle a VM of a Linux server distro.

**Comment by Joly0** [comment]
> If it has enough power to run game servers then it likely can handle a VM of a Linux server distro.

I think the bigger problem is administration overhead. My main job is system administration but even i prefer docker containers over VMs or even LXCs. Setting up a Linux VM or LXC, configure it and in the end keeping it up2date is way more work than "docker pull && docker run", even when cubecoders provides a very strong and good installation script for basically everything it needs. Its mostly just the maintenance that is the hurdle.  Though VMs, as good as they are, still have a bigger resource overhead compared to docker or even LXC.

You are however absolutely correct in that docker host access from a container is inherently insecure, but its still true that many popular management tools (including portainer or komodo) still use this method of deployment, although there are alterantives like docker-socket-proxy.

Also, what i described above for the LXC template is purely meant for Unraid where LXC is not an official supported way, but a developer created a plugin for LXC on unraid which allows to use templates (so pre-defined LXCs basically), which i have used and maintained for a while for AMP
